# Task
Generate Python code for a comprehensive set of list manipulation tasks, including calculating the sum, finding maximum, minimum, second largest, and second smallest elements, removing duplicates, counting element occurrences, reversing and sorting lists without built-in methods, merging lists, adding and deleting elements, counting even, odd, and prime numbers, calculating the product of all elements, printing unique elements, finding common elements and differences between two lists, and rotating lists left and right. Finally, summarize the functionalities of all generated code snippets.

## Input List and Print Sum

### Subtask:
Generate Python code to take a list of numbers as input and calculate their sum.


**Reasoning**:
The subtask requires generating Python code to take a list of numbers as input and calculate their sum. I will use a code block to initialize a list, calculate its sum using the built-in `sum()` function, and then print the result with a descriptive message.



In [1]:
numbers = [10, 20, 30, 40, 50]
total_sum = sum(numbers)
print(f"The sum of the numbers in the list is: {total_sum}")

The sum of the numbers in the list is: 150


## Find Maximum in List

### Subtask:
Generate Python code to find the maximum number in a given list.


**Reasoning**:
The subtask requires finding the maximum number in a list. I will initialize a sample list, use the built-in `max()` function to find the maximum, and then print the result with a descriptive message.



In [2]:
numbers = [10, 20, 5, 40, 30]
max_number = max(numbers)
print(f"The maximum number in the list is: {max_number}")

The maximum number in the list is: 40


## Find Minimum in List

### Subtask:
Generate Python code to find the minimum number in a given list.


**Reasoning**:
The subtask requires finding the minimum number in a given list. I will initialize a sample list, use the built-in `min()` function to find the minimum, and then print the result with a descriptive message.

**Reasoning**:
The subtask is to generate Python code to find the minimum number in a given list. I will initialize a list, use the `min()` function to find the minimum, and then print the result as instructed.



In [3]:
numbers = [10, 20, 5, 40, 30]
min_number = min(numbers)
print(f"The minimum number in the list is: {min_number}")

The minimum number in the list is: 5


## Find Second Largest in List

### Subtask:
Generate Python code to find the second largest number in a given list.

## Find Second Largest in List

### Subtask:
Generate Python code to find the second largest number in a given list.

## Find Second Largest in List

### Subtask:
Generate Python code to find the second largest number in a given list.

**Reasoning**:
The subtask is to generate Python code to find the second largest number in a given list. I will initialize a sample list, then sort it to easily identify the second largest element. I will also include a check to ensure the list has at least two unique elements before attempting to find the second largest, and remove duplicates to prevent incorrect results if duplicates are present.

**Reasoning**:
I will generate Python code to find the second largest number. First, I will remove duplicates and sort the list in ascending order. Then, I will access the second-to-last element to get the second largest number, ensuring the list has enough unique elements to avoid errors.



In [4]:
numbers = [10, 20, 5, 40, 30, 20]
unique_numbers = sorted(list(set(numbers)))

if len(unique_numbers) >= 2:
    second_largest = unique_numbers[-2]
    print(f"The second largest number in the list is: {second_largest}")
else:
    print("The list does not have a second largest number (requires at least two unique elements).")

The second largest number in the list is: 30


## Find Second Smallest in List

### Subtask:
Generate Python code to find the second smallest number in a given list.

**Reasoning**:
The subtask is to generate Python code to find the second smallest number in a given list. I will initialize a sample list, then remove duplicates and sort the list in ascending order. After sorting, the second smallest number will be at index 1. I will also include a check to ensure the list has at least two unique elements before attempting to find the second smallest, to prevent errors.

**Reasoning**:
The subtask is to find the second smallest number in a given list. I will initialize a list, remove duplicates, sort it, and then access the element at index 1 (second position) to get the second smallest number. I will also include a check for lists with fewer than two unique elements.



In [5]:
numbers = [10, 20, 5, 40, 30, 5]
unique_numbers = sorted(list(set(numbers)))

if len(unique_numbers) >= 2:
    second_smallest = unique_numbers[1]
    print(f"The second smallest number in the list is: {second_smallest}")
else:
    print("The list does not have a second smallest number (requires at least two unique elements).")

The second smallest number in the list is: 10


## Remove Duplicates from List

### Subtask:
Generate Python code to remove duplicate elements from a given list.

## Remove Duplicates from List

### Subtask:
Generate Python code to remove duplicate elements from a given list.

## Remove Duplicates from List

### Subtask:
Generate Python code to remove duplicate elements from a given list.

**Reasoning**:
The subtask is to generate Python code to remove duplicate elements from a given list. I will initialize a sample list with duplicates, convert it to a set to automatically remove duplicates, and then convert the set back into a list. Finally, I will print the new list with duplicates removed.

**Reasoning**:
The subtask is to generate Python code to remove duplicate elements from a given list. I will initialize a sample list with duplicates, convert it to a set to remove duplicates, and then convert it back to a list. Finally, I will print the list with duplicates removed.



In [6]:
numbers = [1, 2, 2, 3, 4, 4, 5]
unique_numbers = list(set(numbers))
print(f"Original list with duplicates: {numbers}")
print(f"List after removing duplicates: {unique_numbers}")

Original list with duplicates: [1, 2, 2, 3, 4, 4, 5]
List after removing duplicates: [1, 2, 3, 4, 5]


## Count Element Occurrences

### Subtask:
Generate Python code to count the occurrences of each element in a given list.

## Count Occurrences of Element

### Subtask:
Generate Python code to count the occurrences of each element in a given list.


**Reasoning**:
The subtask is to count the occurrences of each element in a given list. I will use the `collections.Counter` class, which is an efficient way to achieve this. I will first import `Counter`, then initialize a sample list, and finally use `Counter` to get the counts and print the result.



In [7]:
from collections import Counter

numbers = [1, 2, 2, 3, 4, 4, 4, 5]
element_counts = Counter(numbers)

print(f"Original list: {numbers}")
print(f"Element occurrences: {element_counts}")

Original list: [1, 2, 2, 3, 4, 4, 4, 5]
Element occurrences: Counter({4: 3, 2: 2, 1: 1, 3: 1, 5: 1})


## Reverse List without reverse()

### Subtask:
Generate Python code to reverse a list without using the built-in reverse() method or slicing.


**Reasoning**:
The subtask requires reversing a list without using built-in methods like `reverse()` or slicing. I will initialize a sample list, then create an empty list. I will iterate through the original list from the last element to the first, appending each element to the new empty list. Finally, I will print both the original and the reversed list to demonstrate the outcome.

**Reasoning**:
The subtask requires generating Python code to reverse a list without using built-in methods or slicing. I will implement the logic described in the instructions: initialize a list, loop from the end to the beginning, append elements to a new list, and then print both the original and the new reversed list.



In [8]:
my_list = [1, 2, 3, 4, 5]
reversed_list = []

for i in range(len(my_list) - 1, -1, -1):
    reversed_list.append(my_list[i])

print(f"Original list: {my_list}")
print(f"Reversed list (without using reverse() or slicing): {reversed_list}")

Original list: [1, 2, 3, 4, 5]
Reversed list (without using reverse() or slicing): [5, 4, 3, 2, 1]


## Sort List without sort() or sorted()

### Subtask:
Generate Python code to sort a list in ascending order without using the built-in `sort()` method or `sorted()` function.

## Sort List without sort() or sorted()

### Subtask:
Generate Python code to sort a list in ascending order without using the built-in `sort()` method or `sorted()` function.

**Reasoning**:
The subtask is to generate Python code to sort a list in ascending order without using the built-in `sort()` method or `sorted()` function. I will implement a sorting algorithm, such as Bubble Sort. This involves iterating through the list, comparing adjacent elements, and swapping them if they are in the wrong order. This process will be repeated until the list is fully sorted.

**Reasoning**:
The subtask is to generate Python code to sort a list in ascending order without using the built-in `sort()` method or `sorted()` function. I will implement a sorting algorithm, such as Bubble Sort. This involves iterating through the list, comparing adjacent elements, and swapping them if they are in the wrong order. This process will be repeated until the list is fully sorted.

**Reasoning**:
The subtask is to sort a list without using built-in methods. Based on the previous reasoning, I will generate Python code implementing the Bubble Sort algorithm to achieve this.



In [9]:
my_list = [64, 34, 25, 12, 22, 11, 90]
n = len(my_list)

# Bubble Sort implementation
for i in range(n - 1):
    for j in range(0, n - i - 1):
        if my_list[j] > my_list[j + 1]:
            my_list[j], my_list[j + 1] = my_list[j + 1], my_list[j] # Swap elements

print(f"Original list: {[64, 34, 25, 12, 22, 11, 90]}")
print(f"Sorted list (without using sort() or sorted()): {my_list}")

Original list: [64, 34, 25, 12, 22, 11, 90]
Sorted list (without using sort() or sorted()): [11, 12, 22, 25, 34, 64, 90]


## Merge Two Lists

### Subtask:
Generate Python code to merge two given lists into a single list.

## Merge Two Lists

### Subtask:
Generate Python code to merge two given lists into a single list.

**Reasoning**:
The subtask is to generate Python code to merge two given lists into a single list. I will initialize two sample lists, then use the `+` operator to concatenate them, creating a new merged list. Finally, I will print all three lists with descriptive messages.

**Reasoning**:
The subtask is to generate Python code to merge two given lists into a single list. I will initialize two sample lists, then use the `+` operator to concatenate them, creating a new merged list. Finally, I will print all three lists with descriptive messages.

**Reasoning**:
Based on the reasoning provided, I will now generate the Python code to merge two lists using the `+` operator.



In [10]:
list1 = [1, 2, 3]
list2 = [4, 5, 6]

merged_list = list1 + list2

print(f"List 1: {list1}")
print(f"List 2: {list2}")
print(f"Merged list: {merged_list}")

List 1: [1, 2, 3]
List 2: [4, 5, 6]
Merged list: [1, 2, 3, 4, 5, 6]


## Add and Delete Elements

### Subtask:
Generate Python code to add elements to and delete elements from a given list.

**Reasoning**:
The subtask is to generate Python code to add and delete elements from a given list. I will initialize a sample list. For adding elements, I will use the `append()` method to add an element to the end of the list and the `insert()` method to add an element at a specific index. For deleting elements, I will use the `remove()` method to delete a specific value and the `pop()` method to delete an element by its index.

**Reasoning**:
The subtask is to generate Python code to add and delete elements from a given list. I will initialize a sample list. For adding elements, I will use the `append()` method to add an element to the end of the list and the `insert()` method to add an element at a specific index. For deleting elements, I will use the `remove()` method to delete a specific value and the `pop()` method to delete an element by its index.

**Reasoning**:
Based on the reasoning provided, I will now generate the Python code to demonstrate adding and deleting elements from a list using `append()`, `insert()`, `remove()`, and `pop()`.



In [11]:
my_list = [10, 20, 30, 40, 50]
print(f"Original list: {my_list}")

# Adding elements
my_list.append(60) # Add to the end
print(f"After appending 60: {my_list}")

my_list.insert(1, 15) # Insert 15 at index 1
print(f"After inserting 15 at index 1: {my_list}")

# Deleting elements
my_list.remove(30) # Remove by value
print(f"After removing 30: {my_list}")

popped_element = my_list.pop(0) # Remove by index (first element)
print(f"After popping element at index 0 ({popped_element}): {my_list}")

# Demonstrate deleting a specific value that might not be present
# try:
#     my_list.remove(100)
# except ValueError:
#     print("100 not found in list, cannot remove.")


Original list: [10, 20, 30, 40, 50]
After appending 60: [10, 20, 30, 40, 50, 60]
After inserting 15 at index 1: [10, 15, 20, 30, 40, 50, 60]
After removing 30: [10, 15, 20, 40, 50, 60]
After popping element at index 0 (10): [15, 20, 40, 50, 60]


## Count Even Numbers in List

### Subtask:
Generate Python code to count the number of even numbers in a given list.


**Reasoning**:
The subtask is to count the number of even numbers in a given list. I will initialize a sample list of integers and an `even_count` variable to 0. I will then iterate through each number in the list. Inside the loop, I'll use the modulo operator (`%`) to check if the number is even. If it is, I'll increment `even_count`. Finally, I will print both the original list and the total count of even numbers using an f-string.

**Reasoning**:
Based on the reasoning provided, I will now generate the Python code to count the even numbers in a list, following the specified instructions.



In [12]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
even_count = 0

for number in numbers:
    if number % 2 == 0:
        even_count += 1

print(f"Original list: {numbers}")
print(f"Number of even numbers: {even_count}")

Original list: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Number of even numbers: 5


## Count Odd Numbers in List

### Subtask:
Generate Python code to count the number of odd numbers in a given list.

**Reasoning**:
The subtask is to count the number of odd numbers in a given list. I will initialize a sample list of integers and an `odd_count` variable to 0. I will then iterate through each number in the list. Inside the loop, I'll use the modulo operator (`%`) to check if the number is odd (i.e., `number % 2 != 0`). If it is, I'll increment `odd_count`. Finally, I will print both the original list and the total count of odd numbers using an f-string.

**Reasoning**:
The subtask is to count the number of odd numbers in a given list. I will initialize a sample list of integers and an `odd_count` variable to 0. I will then iterate through each number in the list. Inside the loop, I'll use the modulo operator (`%`) to check if the number is odd (i.e., `number % 2 != 0`). If it is, I'll increment `odd_count`. Finally, I will print both the original list and the total count of odd numbers using an f-string.

**Reasoning**:
The subtask is to count the number of odd numbers in a given list. I will initialize a sample list and an odd counter. Then, I will iterate through each number, check if it's odd using the modulo operator, and increment the counter accordingly. Finally, I will print the result.



In [13]:
**Reasoning**:
The subtask is to count the number of odd numbers in a given list. I will initialize a sample list of integers and an `odd_count` variable to 0. I will then iterate through each number in the list. Inside the loop, I'll use the modulo operator (`%`) to check if the number is odd (i.e., `number % 2 != 0`). If it is, I'll increment `odd_count`. Finally, I will print both the original list and the total count of odd numbers using an f-string.

SyntaxError: invalid syntax (ipython-input-3874140099.py, line 1)

**Reasoning**:
The previous code block failed because it incorrectly contained markdown text within a Python code cell. I will now provide the correct Python code to count odd numbers, based on the established reasoning.



In [14]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
odd_count = 0

for number in numbers:
    if number % 2 != 0:
        odd_count += 1

print(f"Original list: {numbers}")
print(f"Number of odd numbers: {odd_count}")

Original list: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Number of odd numbers: 5


## Count Prime Numbers in List

### Subtask:
Generate Python code to count the number of prime numbers in a given list.

## Count Prime Numbers in List

### Subtask:
Generate Python code to count the number of prime numbers in a given list.

**Reasoning**:
The subtask is to count the number of prime numbers in a given list. I will initialize a sample list of integers and a `prime_count` variable to 0. I will define a helper function `is_prime(num)` that returns `True` if `num` is prime and `False` otherwise. This function will handle edge cases like numbers less than or equal to 1 and efficiently check divisibility up to the square root of `num`. Then, I will iterate through each number in the list, call `is_prime()` for each number, and if it's prime, increment `prime_count`. Finally, I will print both the original list and the total count of prime numbers.

**Reasoning**:
Based on the reasoning provided, I will now generate the Python code to count prime numbers in a list. This will involve defining an `is_prime` helper function and then iterating through the sample list to apply this function and tally the prime numbers.



In [15]:
import math

def is_prime(num):
    if num <= 1:
        return False
    for i in range(2, int(math.sqrt(num)) + 1):
        if num % i == 0:
            return False
    return True

numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 17, 19, 23]
prime_count = 0
prime_numbers_list = []

for number in numbers:
    if is_prime(number):
        prime_count += 1
        prime_numbers_list.append(number)

print(f"Original list: {numbers}")
print(f"Number of prime numbers: {prime_count}")
print(f"Prime numbers found: {prime_numbers_list}")

Original list: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 17, 19, 23]
Number of prime numbers: 9
Prime numbers found: [2, 3, 5, 7, 11, 13, 17, 19, 23]


## Calculate Product of All Elements

### Subtask:
Generate Python code to calculate the product of all elements in a given list.

## Calculate Product of All Elements

### Subtask:
Generate Python code to calculate the product of all elements in a given list.

**Reasoning**:
The subtask is to generate Python code to calculate the product of all elements in a given list. I will initialize a sample list of numbers and a `product` variable to 1. I will then iterate through each number in the list, multiplying it with the `product` variable. Finally, I will print both the original list and the calculated product using an f-string.

**Reasoning**:
The subtask is to generate Python code to calculate the product of all elements in a given list. I will initialize a sample list of numbers and a `product` variable to 1. I will then iterate through each number in the list, multiplying it with the `product` variable. Finally, I will print both the original list and the calculated product using an f-string.

**Reasoning**:
Based on the reasoning provided, I will now generate the Python code to calculate the product of all elements in a list, following the specified instructions.



In [16]:
numbers = [1, 2, 3, 4, 5]
product = 1

for number in numbers:
    product *= number

print(f"Original list: {numbers}")
print(f"Product of all elements: {product}")

Original list: [1, 2, 3, 4, 5]
Product of all elements: 120


## Print Unique Elements

### Subtask:
Generate Python code to print only the unique elements from a list.


**Reasoning**:
The subtask is to generate Python code to print only the unique elements from a list. I will initialize a sample list with duplicate elements, convert it to a `set` to automatically remove duplicates, and then convert the `set` back into a `list` for printing. Finally, I will print both the original list and the list of unique elements with descriptive labels.

**Reasoning**:
Based on the reasoning provided, I will now generate the Python code to print only the unique elements from a list, following the specified instructions.



In [17]:
my_list = [1, 2, 2, 3, 4, 4, 5, 1]

# Convert to a set to remove duplicates
unique_elements_set = set(my_list)

# Convert back to a list (optional, for ordered output or if a list is strictly required)
unique_elements_list = sorted(list(unique_elements_set))

print(f"Original list: {my_list}")
print(f"Unique elements (as set): {unique_elements_set}")
print(f"Unique elements (as sorted list): {unique_elements_list}")

Original list: [1, 2, 2, 3, 4, 4, 5, 1]
Unique elements (as set): {1, 2, 3, 4, 5}
Unique elements (as sorted list): [1, 2, 3, 4, 5]


## Find Common Elements and Differences between Two Lists

### Subtask:
Generate Python code to find the common elements and the elements unique to each of two given lists.

**Reasoning**:
The subtask is to find the common elements and the differences between two given lists. I will initialize two sample lists. To find common elements, I will convert both lists to sets and use the `intersection()` method. To find elements unique to each list, I will use the `difference()` method on each set. Finally, I will print the original lists, the common elements, and the unique elements from each list with descriptive labels.

**Reasoning**:
The subtask is to find the common elements and the differences between two given lists. I will initialize two sample lists. To find common elements, I will convert both lists to sets and use the `intersection()` method. To find elements unique to each list, I will use the `difference()` method on each set. Finally, I will print the original lists, the common elements, and the unique elements from each list with descriptive labels.

**Reasoning**:
Based on the reasoning provided, I will now generate the Python code to find the common elements and differences between two lists using set operations.



In [18]:
list1 = [1, 2, 3, 4, 5]
list2 = [4, 5, 6, 7, 8]

set1 = set(list1)
set2 = set(list2)

common_elements = list(set1.intersection(set2))
elements_only_in_list1 = list(set1.difference(set2))
elements_only_in_list2 = list(set2.difference(set1))

print(f"List 1: {list1}")
print(f"List 2: {list2}")
print(f"Common elements: {common_elements}")
print(f"Elements only in List 1: {elements_only_in_list1}")
print(f"Elements only in List 2: {elements_only_in_list2}")

List 1: [1, 2, 3, 4, 5]
List 2: [4, 5, 6, 7, 8]
Common elements: [4, 5]
Elements only in List 1: [1, 2, 3]
Elements only in List 2: [8, 6, 7]


## Rotate Lists Left and Right

### Subtask:
Generate Python code to rotate a given list to the left and to the right by a specified number of positions.

**Reasoning**:
The subtask is to generate Python code to rotate a given list to the left and to the right by a specified number of positions. I will initialize a sample list and define two functions: `rotate_left(lst, k)` and `rotate_right(lst, k)`. Both functions will handle positive and negative rotation values and ensure that rotations wrap around the list length. The `rotate_left` function will shift elements to the left, and the `rotate_right` function will shift elements to the right. Finally, I will demonstrate both rotations with examples and print the results with descriptive labels.

**Reasoning**:
Based on the reasoning provided, I will now generate the Python code to rotate a list to the left and to the right by a specified number of positions, implementing the `rotate_left` and `rotate_right` functions.



In [19]:
def rotate_left(lst, k):
    n = len(lst)
    if n == 0: # Handle empty list
        return []
    k = k % n # Ensure k is within list bounds
    return lst[k:] + lst[:k]

def rotate_right(lst, k):
    n = len(lst)
    if n == 0: # Handle empty list
        return []
    k = k % n # Ensure k is within list bounds
    return lst[n - k:] + lst[:n - k]

my_list = [1, 2, 3, 4, 5]
positions = 2

print(f"Original list: {my_list}")

rotated_left = rotate_left(my_list, positions)
print(f"List rotated left by {positions} positions: {rotated_left}")

rotated_right = rotate_right(my_list, positions)
print(f"List rotated right by {positions} positions: {rotated_right}")


Original list: [1, 2, 3, 4, 5]
List rotated left by 2 positions: [3, 4, 5, 1, 2]
List rotated right by 2 positions: [4, 5, 1, 2, 3]


## Summary of List Manipulation Functionalities

This notebook demonstrates various Python list manipulation techniques, covering a comprehensive set of operations:

1.  **Calculate Sum**: Computes the sum of all elements in a list.
2.  **Find Maximum**: Identifies the largest element in a list.
3.  **Find Minimum**: Identifies the smallest element in a list.
4.  **Find Second Largest**: Locates the second largest unique element in a list after removing duplicates and sorting.
5.  **Find Second Smallest**: Locates the second smallest unique element in a list after removing duplicates and sorting.
6.  **Remove Duplicates**: Creates a new list containing only the unique elements from an original list.
7.  **Count Element Occurrences**: Determines how many times each element appears in a list using `collections.Counter`.
8.  **Reverse List (without built-in methods)**: Reverses the order of elements in a list using a loop, avoiding `reverse()` or slicing.
9.  **Sort List (without built-in methods)**: Sorts a list in ascending order using the Bubble Sort algorithm, avoiding `sort()` or `sorted()`.
10. **Merge Two Lists**: Combines two separate lists into a single new list.
11. **Add and Delete Elements**: Demonstrates adding elements using `append()` and `insert()`, and deleting elements using `remove()` (by value) and `pop()` (by index).
12. **Count Even Numbers**: Counts the total number of even integers within a list.
13. **Count Odd Numbers**: Counts the total number of odd integers within a list.
14. **Count Prime Numbers**: Identifies and counts prime numbers within a list using a helper function.
15. **Calculate Product of All Elements**: Computes the product of all numbers in a list.
16. **Print Unique Elements**: Extracts and displays only the unique elements from a list.
17. **Find Common Elements and Differences**: Determines elements present in both lists (common) and elements unique to each list using set operations.
18. **Rotate Lists Left and Right**: Implements functions to rotate a list's elements by a specified number of positions to the left or right.

## Summary of List Manipulation Functionalities

This notebook demonstrates various Python list manipulation techniques, covering a comprehensive set of operations:

1.  **Calculate Sum**: Computes the sum of all elements in a list.
2.  **Find Maximum**: Identifies the largest element in a list.
3.  **Find Minimum**: Identifies the smallest element in a list.
4.  **Find Second Largest**: Locates the second largest unique element in a list after removing duplicates and sorting.
5.  **Find Second Smallest**: Locates the second smallest unique element in a list after removing duplicates and sorting.
6.  **Remove Duplicates**: Creates a new list containing only the unique elements from an original list.
7.  **Count Element Occurrences**: Determines how many times each element appears in a list using `collections.Counter`.
8.  **Reverse List (without built-in methods)**: Reverses the order of elements in a list using a loop, avoiding `reverse()` or slicing.
9.  **Sort List (without built-in methods)**: Sorts a list in ascending order using the Bubble Sort algorithm, avoiding `sort()` or `sorted()`.
10. **Merge Two Lists**: Combines two separate lists into a single new list.
11. **Add and Delete Elements**: Demonstrates adding elements using `append()` and `insert()`, and deleting elements using `remove()` (by value) and `pop()` (by index).
12. **Count Even Numbers**: Counts the total number of even integers within a list.
13. **Count Odd Numbers**: Counts the total number of odd integers within a list.
14. **Count Prime Numbers**: Identifies and counts prime numbers within a list using a helper function.
15. **Calculate Product of All Elements**: Computes the product of all numbers in a list.
16. **Print Unique Elements**: Extracts and displays only the unique elements from a list.
17. **Find Common Elements and Differences**: Determines elements present in both lists (common) and elements unique to each list using set operations.
18. **Rotate Lists Left and Right**: Implements functions to rotate a list's elements by a specified number of positions to the left or right.

## Summary of List Manipulation Functionalities

This notebook demonstrates various Python list manipulation techniques, covering a comprehensive set of operations:

1.  **Calculate Sum**: Computes the sum of all elements in a list.
2.  **Find Maximum**: Identifies the largest element in a list.
3.  **Find Minimum**: Identifies the smallest element in a list.
4.  **Find Second Largest**: Locates the second largest unique element in a list after removing duplicates and sorting.
5.  **Find Second Smallest**: Locates the second smallest unique element in a list after removing duplicates and sorting.
6.  **Remove Duplicates**: Creates a new list containing only the unique elements from an original list.
7.  **Count Element Occurrences**: Determines how many times each element appears in a list using `collections.Counter`.
8.  **Reverse List (without built-in methods)**: Reverses the order of elements in a list using a loop, avoiding `reverse()` or slicing.
9.  **Sort List (without built-in methods)**: Sorts a list in ascending order using the Bubble Sort algorithm, avoiding `sort()` or `sorted()`.
10. **Merge Two Lists**: Combines two separate lists into a single new list.
11. **Add and Delete Elements**: Demonstrates adding elements using `append()` and `insert()`, and deleting elements using `remove()` (by value) and `pop()` (by index).
12. **Count Even Numbers**: Counts the total number of even integers within a list.
13. **Count Odd Numbers**: Counts the total number of odd integers within a list.
14. **Count Prime Numbers**: Identifies and counts prime numbers within a list using a helper function.
15. **Calculate Product of All Elements**: Computes the product of all numbers in a list.
16. **Print Unique Elements**: Extracts and displays only the unique elements from a list.
17. **Find Common Elements and Differences**: Determines elements present in both lists (common) and elements unique to each list using set operations.
18. **Rotate Lists Left and Right**: Implements functions to rotate a list's elements by a specified number of positions to the left or right.

## Summary of List Manipulation Functionalities

This notebook demonstrates various Python list manipulation techniques, covering a comprehensive set of operations:

1.  **Calculate Sum**: Computes the sum of all elements in a list.
2.  **Find Maximum**: Identifies the largest element in a list.
3.  **Find Minimum**: Identifies the smallest element in a list.
4.  **Find Second Largest**: Locates the second largest unique element in a list after removing duplicates and sorting.
5.  **Find Second Smallest**: Locates the second smallest unique element in a list after removing duplicates and sorting.
6.  **Remove Duplicates**: Creates a new list containing only the unique elements from an original list.
7.  **Count Element Occurrences**: Determines how many times each element appears in a list using `collections.Counter`.
8.  **Reverse List (without built-in methods)**: Reverses the order of elements in a list using a loop, avoiding `reverse()` or slicing.
9.  **Sort List (without built-in methods)**: Sorts a list in ascending order using the Bubble Sort algorithm, avoiding `sort()` or `sorted()`.
10. **Merge Two Lists**: Combines two separate lists into a single new list.
11. **Add and Delete Elements**: Demonstrates adding elements using `append()` and `insert()`, and deleting elements using `remove()` (by value) and `pop()` (by index).
12. **Count Even Numbers**: Counts the total number of even integers within a list.
13. **Count Odd Numbers**: Counts the total number of odd integers within a list.
14. **Count Prime Numbers**: Identifies and counts prime numbers within a list using a helper function.
15. **Calculate Product of All Elements**: Computes the product of all numbers in a list.
16. **Print Unique Elements**: Extracts and displays only the unique elements from a list.
17. **Find Common Elements and Differences**: Determines elements present in both lists (common) and elements unique to each list using set operations.
18. **Rotate Lists Left and Right**: Implements functions to rotate a list's elements by a specified number of positions to the left or right.

## Summary of List Manipulation Functionalities

This notebook demonstrates various Python list manipulation techniques, covering a comprehensive set of operations:

1.  **Calculate Sum**: Computes the sum of all elements in a list.
2.  **Find Maximum**: Identifies the largest element in a list.
3.  **Find Minimum**: Identifies the smallest element in a list.
4.  **Find Second Largest**: Locates the second largest unique element in a list after removing duplicates and sorting.
5.  **Find Second Smallest**: Locates the second smallest unique element in a list after removing duplicates and sorting.
6.  **Remove Duplicates**: Creates a new list containing only the unique elements from an original list.
7.  **Count Element Occurrences**: Determines how many times each element appears in a list using `collections.Counter`.
8.  **Reverse List (without built-in methods)**: Reverses the order of elements in a list using a loop, avoiding `reverse()` or slicing.
9.  **Sort List (without built-in methods)**: Sorts a list in ascending order using the Bubble Sort algorithm, avoiding `sort()` or `sorted()`.
10. **Merge Two Lists**: Combines two separate lists into a single new list.
11. **Add and Delete Elements**: Demonstrates adding elements using `append()` and `insert()`, and deleting elements using `remove()` (by value) and `pop()` (by index).
12. **Count Even Numbers**: Counts the total number of even integers within a list.
13. **Count Odd Numbers**: Counts the total number of odd integers within a list.
14. **Count Prime Numbers**: Identifies and counts prime numbers within a list using a helper function.
15. **Calculate Product of All Elements**: Computes the product of all numbers in a list.
16. **Print Unique Elements**: Extracts and displays only the unique elements from a list.
17. **Find Common Elements and Differences**: Determines elements present in both lists (common) and elements unique to each list using set operations.
18. **Rotate Lists Left and Right**: Implements functions to rotate a list's elements by a specified number of positions to the left or right.

## Summary of List Manipulation Functionalities

This notebook demonstrates various Python list manipulation techniques, covering a comprehensive set of operations:

1.  **Calculate Sum**: Computes the sum of all elements in a list.
2.  **Find Maximum**: Identifies the largest element in a list.
3.  **Find Minimum**: Identifies the smallest element in a list.
4.  **Find Second Largest**: Locates the second largest unique element in a list after removing duplicates and sorting.
5.  **Find Second Smallest**: Locates the second smallest unique element in a list after removing duplicates and sorting.
6.  **Remove Duplicates**: Creates a new list containing only the unique elements from an original list.
7.  **Count Element Occurrences**: Determines how many times each element appears in a list using `collections.Counter`.
8.  **Reverse List (without built-in methods)**: Reverses the order of elements in a list using a loop, avoiding `reverse()` or slicing.
9.  **Sort List (without built-in methods)**: Sorts a list in ascending order using the Bubble Sort algorithm, avoiding `sort()` or `sorted()`.
10. **Merge Two Lists**: Combines two separate lists into a single new list.
11. **Add and Delete Elements**: Demonstrates adding elements using `append()` and `insert()`, and deleting elements using `remove()` (by value) and `pop()` (by index).
12. **Count Even Numbers**: Counts the total number of even integers within a list.
13. **Count Odd Numbers**: Counts the total number of odd integers within a list.
14. **Count Prime Numbers**: Identifies and counts prime numbers within a list using a helper function.
15. **Calculate Product of All Elements**: Computes the product of all numbers in a list.
16. **Print Unique Elements**: Extracts and displays only the unique elements from a list.
17. **Find Common Elements and Differences**: Determines elements present in both lists (common) and elements unique to each list using set operations.
18. **Rotate Lists Left and Right**: Implements functions to rotate a list's elements by a specified number of positions to the left or right.

## Summary of List Manipulation Functionalities

This notebook demonstrates various Python list manipulation techniques, covering a comprehensive set of operations:

1.  **Calculate Sum**: Computes the sum of all elements in a list.
2.  **Find Maximum**: Identifies the largest element in a list.
3.  **Find Minimum**: Identifies the smallest element in a list.
4.  **Find Second Largest**: Locates the second largest unique element in a list after removing duplicates and sorting.
5.  **Find Second Smallest**: Locates the second smallest unique element in a list after removing duplicates and sorting.
6.  **Remove Duplicates**: Creates a new list containing only the unique elements from an original list.
7.  **Count Element Occurrences**: Determines how many times each element appears in a list using `collections.Counter`.
8.  **Reverse List (without built-in methods)**: Reverses the order of elements in a list using a loop, avoiding `reverse()` or slicing.
9.  **Sort List (without built-in methods)**: Sorts a list in ascending order using the Bubble Sort algorithm, avoiding `sort()` or `sorted()`.
10. **Merge Two Lists**: Combines two separate lists into a single new list.
11. **Add and Delete Elements**: Demonstrates adding elements using `append()` and `insert()`, and deleting elements using `remove()` (by value) and `pop()` (by index).
12. **Count Even Numbers**: Counts the total number of even integers within a list.
13. **Count Odd Numbers**: Counts the total number of odd integers within a list.
14. **Count Prime Numbers**: Identifies and counts prime numbers within a list using a helper function.
15. **Calculate Product of All Elements**: Computes the product of all numbers in a list.
16. **Print Unique Elements**: Extracts and displays only the unique elements from a list.
17. **Find Common Elements and Differences**: Determines elements present in both lists (common) and elements unique to each list using set operations.
18. **Rotate Lists Left and Right**: Implements functions to rotate a list's elements by a specified number of positions to the left or right.

## Final Task

### Subtask:
Summarize the generated code snippets and their functionalities for list manipulations.


## Summary:

### Data Analysis Key Findings

The agent successfully generated and executed Python code for a comprehensive set of 18 list manipulation tasks, demonstrating diverse functionalities:

*   **Basic Operations**:
    *   Calculated the sum of a list (e.g., `[10, 20, 30, 40, 50]` sums to `150`).
    *   Found the maximum element (e.g., `40` in `[10, 20, 5, 40, 30]`).
    *   Found the minimum element (e.g., `5` in `[10, 20, 5, 40, 30]`).
*   **Advanced Element Retrieval**:
    *   Identified the second largest unique element (e.g., `30` from unique sorted `[5, 10, 20, 30, 40]` of `[10, 20, 5, 40, 30, 20]`).
    *   Identified the second smallest unique element (e.g., `10` from unique sorted `[5, 10, 20, 30, 40]` of `[10, 20, 5, 40, 30, 5]`).
*   **List Modification & Cleaning**:
    *   Removed duplicates from a list (e.g., `[1, 2, 2, 3, 4, 4, 5]` became `[1, 2, 3, 4, 5]`).
    *   Counted occurrences of each element (e.g., `[1, 2, 2, 3, 4, 4, 4, 5]` resulted in `Counter({4: 3, 2: 2, 1: 1, 3: 1, 5: 1})`).
    *   Reversed a list without built-in methods or slicing (e.g., `[1, 2, 3, 4, 5]` became `[5, 4, 3, 2, 1]`).
    *   Sorted a list without built-in `sort()` or `sorted()` using Bubble Sort (e.g., `[64, 34, 25, 12, 22, 11, 90]` became `[11, 12, 22, 25, 34, 64, 90]`).
    *   Merged two lists (e.g., `[1, 2, 3]` and `[4, 5, 6]` merged to `[1, 2, 3, 4, 5, 6]`).
    *   Demonstrated adding elements using `append()` and `insert()`, and deleting elements using `remove()` and `pop()` from a list (e.g., `[10, 20, 30, 40, 50]` modified to `[15, 20, 40, 50, 60]` after several operations).
*   **Conditional Counting**:
    *   Counted even numbers (e.g., `5` even numbers in `[1, ..., 10]`).
    *   Counted odd numbers (e.g., `5` odd numbers in `[1, ..., 10]`).
    *   Counted prime numbers using a helper function (e.g., `9` prime numbers in `[1, ..., 23]`).
*   **List Transformations & Comparisons**:
    *   Calculated the product of all elements (e.g., `[1, 2, 3, 4, 5]` has a product of `120`).
    *   Printed unique elements from a list (e.g., `[1, 2, 3, 4, 5]` from `[1, 2, 2, 3, 4, 4, 5, 1]`).
    *   Found common elements (e.g., `[4, 5]`) and differences (e.g., `[1, 2, 3]` and `[8, 6, 7]`) between two lists `[1, 2, 3, 4, 5]` and `[4, 5, 6, 7, 8]`.
    *   Rotated lists left and right by a specified number of positions (e.g., `[1, 2, 3, 4, 5]` rotated left by 2 became `[3, 4, 5, 1, 2]`, and rotated right by 2 became `[4, 5, 1, 2, 3]`).

### Insights or Next Steps

*   The comprehensive set of list manipulation techniques demonstrates proficiency in Python's built-in functions, `collections.Counter`, set operations, and custom algorithm implementation (e.g., Bubble Sort, manual list reversal, prime number check, list rotation).
*   For improved efficiency, especially with very large lists, consider exploring more optimized algorithms for sorting (e.g., Merge Sort, Quick Sort) or for finding unique/common elements (if the order needs to be preserved and set conversion isn't suitable).
